### ISSUE TITLE
Handle NaNs in flow shape factors

### ISSUE THREAD

**Issue by ma-sadeghi** [PMEAL/OpenPNM#1463]
As Zohaib and I recently discussed, for some images, SNOW algorithm produces throats that are larger than the adjacent pores. Since in `OpenPNM`, `ball_and_stick` is the default pore-scale model for flow shape factors, such scenario leads to NaNs in the respective conduits. We should manually handle these NaNs and throw a critical warning so the user is aware that this is happening behind the scenes.

**Comment by ma-sadeghi** [comment]
Actually, we just realized that `ball_and_stick` handles the special case where throat diameter is greater than pore diameter. The problem seems to come from the section in `throat.endpoints` model:
```
    try:
        TC = network[throat_centroid][throats]
        LP1T = _sp.linalg.norm(TC - xyz[cn[:, 0]], axis=1) + 1e-15
        LP2T = _sp.linalg.norm(TC - xyz[cn[:, 1]], axis=1) + 1e-15
        unit_vec_P1T = (TC - xyz[cn[:, 0]]) / LP1T[:, None]
        unit_vec_P2T = (TC - xyz[cn[:, 1]]) / LP2T[:, None]
```
which handles the case where `throat.centroid` is available. Commenting out this part solves the problem.

**Comment by ma-sadeghi** [comment]
Ok, it's not the fault of `throat.centroid`. It's actually the final part of the model:
```
    # Handle throats w/ overlapping pores
    L1 = (4*L**2 + D1**2 - D2**2) / (8*L)
    # L2 = (4*L**2 + D2**2 - D1**2) / (8*L)
    h = (2*_np.sqrt(D1**2/4 - L1**2)).real
    overlap = L - 0.5 * (D1+D2) < 0
    mask = overlap & (Dt < h)
    EP1[mask] = EP2[mask] = (xyz[cn[:, 0]] + L1[:, None] * unit_vec_P1T)[mask]
```
which handles the case of overlapping pores. Basically, the line where it says `EP1[mask] = EP2[mask] = ...` assumes that the two endpoints coincide, which is true when pore centroids and throat centroid are co-linear, but definitely not true otherwise. See the diagram below:
![Untitled](https://user-images.githubusercontent.com/14086031/79081413-29e12080-7ceb-11ea-811b-a31d7285d679.png)
Basically, by assuming `endpoint.head` and `endpoint.tail` are identical, we've saved ourselves 1 extra line of code:
```
EP2[mask] = (xyz[cn[:, 1]] + L2[:, None] * unit_vec_P2T)[mask]
```
which cost us this nasty bug! So basically, we need to replace
```
EP1[mask] = EP2[mask] = (xyz[cn[:, 0]] + L1[:, None] * unit_vec_P1T)[mask]
```
with
```
EP1[mask] = (xyz[cn[:, 0]] + L1[:, None] * unit_vec_P1T)[mask]
EP2[mask] = (xyz[cn[:, 1]] + L2[:, None] * unit_vec_P2T)[mask]
```
Although we've just released 2.3, I think this bug is critical enough to merit releasing a hotfix.
